fix 500 error when course related to a course run is changed manually

DawoudSheraz opened the pull request:
### PROD-137

### Description
Whenever a course run is published from Publisher, its equivalent is created in the course metadata(and other locations). The course metadata, along with other places, is supposed to be in sync with the publisher. It is possible to cause the data inconsistency in the course metadata from the discovery Django admin. One such case is the changing of the parent course of a course run in the metadata. This inconsistency leads to the inability of the user to publish the course run again as the unexpected behavior is not what is to be expected. Course teams do come across a situation where they have to change the parent course for a course run in Discovery.

This PR creates a patch to allow for publish to work even after causing the data inconsistency. This PR assumes that Discovery course has precedence over publisher course and if a discover course run already has an associated course, use it instead of publisher course run's course

DawoudSheraz commented:
This fix reverts the course change done in the course metadata as all the course creation in the metadata is done using the publisher course run and the changes are not 2-way i.e. only publisher changes are supposed to reflect in both publisher & Metadata.

awaisdar001 commented:
Are going to add a test?

staubina commented:
I am going to review this.  I have also asked @mikix to take a look.

staubina commented on course_discovery/apps/publisher/api/v1/views.py:
```diff
@@ -155,8 +155,13 @@ def publish_to_discovery(self, partner, course_run):
+            'has_ofac_restrictions': course_run.has_ofac_restrictions,
         }
+
+        discovery_course_run = DiscoveryCourseRun.objects.filter(key=course_run.lms_course_id).first()
```
This seems like a sensible approach to me.

This should no longer be an issue as we move to the new version of Publisher we are working on that will not rely on the Publisher data model.

mikix commented on course_discovery/apps/publisher/api/v1/views.py:
```diff
@@ -155,8 +155,13 @@ def publish_to_discovery(self, partner, course_run):
+            'has_ofac_restrictions': course_run.has_ofac_restrictions,
         }
+
+        discovery_course_run = DiscoveryCourseRun.objects.filter(key=course_run.lms_course_id).first()
```
You can rewrite this as `discovery_course_run = course_run.discovery_course_run` - there's a dynamic property that does that query for you, with an early exit for None.

Actually, it may be worth adding a new dynamic property called `discovery_course` on the Publisher.CourseRun model, that would do this whole lookup you're doing - to encourage others to avoid this bug? Maybe even with some nice comments about why the property exists.

(And regardless if you do above - some nice comments here about why you're doing these lines might be nice - would be confusing if you don't have context.)

DawoudSheraz replied:
I don't know about the dynamic property, that's why I did it like this. I will take a look at it. and I will add a comment later. It was just a quick PR update to see if the fix would work.

DawoudSheraz replied:
I looked at the mentioned property, and it is doing the same as I am doing at this location. Thanks for pointing that out. and I will try adding `discovery_course` property as it does look worth an effort.

mikix commented on course_discovery/apps/publisher/api/v1/views.py:
```diff
@@ -155,8 +155,13 @@ def publish_to_discovery(self, partner, course_run):
+
+        discovery_course_run = DiscoveryCourseRun.objects.filter(key=course_run.lms_course_id).first()
+        if discovery_course_run and discovery_course_run.course:
+            discovery_course = discovery_course_run.course
```
So you're overriding the `discovery_course` object from earlier in this method (where one was updated_or_created). Why not do this check earlier and update the existing CourseMetadata.Course object if it exists? Rather than update the wrong one, then finally grabbing the right one here?

DawoudSheraz replied:
That is a twilight zone as if course team changed the course, that doesn't necessarily mean that the previous course was entirely wrong. I have recently investigated the case where the course team changed course because there were enrollments in the old course and new course didn't have any. If I were to do so as you are trying to ask, it might cause more data inconsistencies. But I can try investigating and analyze the results.

mikix commented on course_discovery/apps/publisher/api/v1/tests/test_views.py:
```diff
@@ -225,6 +225,64 @@ def test_publish(self, mock_access_token):  # pylint: disable=unused-argument,to
 
+    @responses.activate
+    @mock.patch.object(Partner, 'access_token', return_value='JWT fake')
+    @override_switch(PUBLISHER_ENABLE_READ_ONLY_FIELDS, active=True)
```
We are actively planning to remove this switch. It would make our future lives easier if we didn't add more tests that rely on it. Could maybe something else be mocked to avoid whatever you're trying to avoid here?

DawoudSheraz replied:
Sure.

DawoudSheraz commented:
@staubina @mikix There are some questions that I would like you guys to help in answering. First, the place where I have introduced the fix, after that place, the `discovery_course` is used for two purposes:
1. Course [Entitlements](https://github.com/edx/course-discovery/blob/b6a535826ac928a740e2e9f0dbb0d34c0834c7c7/course_discovery/apps/publisher/api/v1/views.py#L175)
2. If created is true, then [associate](https://github.com/edx/course-discovery/blob/b6a535826ac928a740e2e9f0dbb0d34c0834c7c7/course_discovery/apps/publisher/api/v1/views.py#L205) the newly created `discovery_course_run` with `discovery_course`.

Since we have changed the parent course ourselves, are the entitlements supposed to be created with the changed course or with publisher course? Also, Is there any effect when the `discovery_course_run` is made canoncial_course_run of `discovery_course`?

I would also like to address again the question that @mikix asked:

> So you're overriding the discovery_course object from earlier in this method (where one was updated_or_created). Why not do this check earlier and update the existing CourseMetadata.Course object if it exists? Rather than update the wrong one, then finally grabbing the right one here?

The problem is that If we try to make the `discovery_course` consistent across the function, it causes more problems. I will explain it with an example. Say that for discovery run, we originally had the course 001. We changed it to 002 manually. In Publisher, it is still 001. In terms of the data, 001 has subjects s1,s2 and s3 and 002 has None. If I make `discovery_course` consistent by moving it to top of the function and avoid [this](https://github.com/edx/course-discovery/blob/b6a535826ac928a740e2e9f0dbb0d34c0834c7c7/course_discovery/apps/publisher/api/v1/views.py#L133) call, after publish, now 002 will also contain s1, s2 & s3. I am sure this is not something we would want to do as the data inconsistency will grow tremendously. If we want to create the Entitlements with the changed course, then my fix will not cause any issues. But, yes, the `discovery_course` object will be inconsistent in first half of the function, and I can't think of any other way at the moment to avoid that situation.

mikix commented:
I see what you're saying about the inconsistent courses. I hope/assume that people in this situation will know the score and will not expect edits to the wrong course info in publisher to show up anywhere useful.

And I would say that if we take that stance - that course-level edits on this abandoned course are ignored - we don't have to worry about setting entitlement info either, since that's a course-level thing. (So continue to set entitlements before looking up new discovery_course.) Does that seem correct?

staubina commented:
The Entitlement should be created with the Course that the Course Run is pointing to in Discovery Course metadata.  This is important because when the entitlement is used it will use the Course UUID to get the available Course Runs.

The Course UUID is also sent to ecommerce.

DawoudSheraz commented:
@staubina @mikix I discussed this problem within the team and we have reached the conclusion that the fix will cause problems for E-commerce. When publishing to e-commerce, the Discovery course is being fetched using partner and course_key([reference](https://github.com/edx/course-discovery/blob/6f9e323a7a3d7e5a36c8123f693947162a4ffab7/course_discovery/apps/publisher/api/v1/views.py#L87)). The course_key is being made from the publisher course, and when the Discovery course is queried, it returns the one that corresponds to publisher course. I tested this locally, and here is how the flow went:

#### Note: I initally had to disable publish_to_studio and publish_to_ecommerce as I haven't set them locally
1. I had a course_run cr_1 in publisher, that corresponded to c_1 in publisher.
2. I published the course and same were created in Metadata.
3. I changed the metadata course run's parent course to c_2.
4. Press publish. The publish happened.
5. I put a breakpoint in publish_to_ecommerce and pressed publish. In ecommerce, it would make the key with publisher course, so it would be like key_c_1. When it tries to get the Discovery course for that key, it is able to find and returns the course with key_c_1, but we would want key_c_2. Since it gets c_1 course, it will pass its UUID and ecommerce data will go out of sync.
6. Also, the [canonical course run](https://github.com/edx/course-discovery/blob/6f9e323a7a3d7e5a36c8123f693947162a4ffab7/course_discovery/apps/publisher/api/v1/views.py#L205) bit doesn't update as we are not making the new course run, so the course run is the canonical course run of c_1, not c_2.

The problem that have arisen is that wherever we are getting Discovery course based on publisher course run, we will have to check and get the discovery course from publisher run. I have added that property, but I am not sure of the scope of the places where the fix will be required.

staubina commented:
@DawoudSheraz Are you planning to continue to investigate and work on this issue?

FYI @fysheets

staubina commented:
Thank you for the detailed status.

DawoudSheraz commented:
@staubina Yes, I would like to continue the work, if possible. Do you have any suggestions for any further action?

staubina commented:
It seems like we should take a step back and assess any other changes that would be required to make this work. 

Then we should assess the value in fixing this here or fixing it in the new version of Publisher which relies only on the Course Metadata database.

> I put a breakpoint in publish_to_ecommerce and pressed publish. In ecommerce, it would make the key with publisher course, so it would be like key_c_1. When it tries to get the Discovery course for that key, it is able to find and returns the course with key_c_1, but we would want key_c_2. Since it gets c_1 course, it will pass its UUID and ecommerce data will go out of sync.

At some point, we send the Course Key and potentially the Course UUID for entitlements to create the products.  We need to make sure we sent the course key or UUID for c_2 in this case.

I would assume this would be wherever we create the Products.

DawoudSheraz commented:
Given all the conversation we had on the PR, this patch is not enough to allow publishing in case of data inconsistencies. Closing this PR.
